### PR TITLE
Rec: proxy by table per domain queried

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Followup on #11396 

It turns out there needs to be another condition on which to let the mapping depend: the name queried.

As a bonus, fix some small issues in the `DNS Suffix Match Group` docs.

ping @neilcook 

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X ] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [X] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/docs/lua-config/proxymapping.rst [hunk 1]
```diff
@@ -39,13 +40,17 @@ If the subnets specified in multiple :func:`addProxyMapping` calls overlap, the
...
 
   .. versionadded:: 4.7.0
 
   Specify a table based mapping for a subnet.
 
   :param string subnet: a subnet to match
   :param string ip: the IP address or IPaddress port combination to match the subnet to.
+  :param array domains: An array of strings used to fill a :ref:`dns-suffix-match-group`.
 
+If this ``domains`` argument is given to this function, only queries for names matching the :ref:`dns-suffix-match-group` will use the value ``M`` to determine the outgoing ECS, other queries will use the value ``S``.
```
Comment (neilcook) [inline pdns/recursordist/docs/lua-config/proxymapping.rst:53]:
```suggestion
If the optional ``domains`` argument is given to this function, only queries for names matching the :ref:`dns-suffix-match-group` will use the value ``M`` to determine the outgoing ECS; other queries will use the value ``S``.
```

File: pdns/rec-lua-conf.cc [hunk 2]
```diff
@@ -725,11 +725,18 @@ void loadRecursorLuaConfig(const std::string& fname, luaConfigDelayedThreads& de
     lci.allowAdditionalQTypes.insert_or_assign(qtype, pair(targets, mode));
   });
 
-  Lua->writeFunction("addProxyMapping", [&proxyMapping](const string& netmaskArg, const string& addressArg) {
+  Lua->writeFunction("addProxyMapping", [&proxyMapping](const string& netmaskArg, const string& addressArg, boost::optional<std::vector<pair<int,std::string>>> smnStrings) {
     try {
       Netmask netmask(netmaskArg);
       ComboAddress address(addressArg);
-      proxyMapping.insert_or_assign(netmask, address);
+      boost::optional<SuffixMatchNode> smn;
```
Comment (Habbie) [inline pdns/rec-lua-conf.cc:732]: Maybe this could be `std::optional`? Not very important, but it'll be one fewer cleanup, some day.

**Review by Habbie — APPROVED** [review]
Few nits, please read them - otherwise, looks good.

File: pdns/rec-lua-conf.cc (hunk 2 above)
Comment (omoerbeek) [reply pdns/rec-lua-conf.cc:732]: This ends up in a class handled by `LuaWrapper`, which has no `std::optional` support afaik.

Comment (Habbie) [reply pdns/rec-lua-conf.cc:732]: Ah, I didn't think that class would also hit `LuaWrapper`. In that case, indeed, carry on.

File: pdns/recursordist/docs/lua-config/proxymapping.rst [hunk 3]
```diff
@@ -39,13 +40,17 @@ If the subnets specified in multiple :func:`addProxyMapping` calls overlap, the
...
 
   .. versionadded:: 4.7.0
 
   Specify a table based mapping for a subnet.
 
   :param string subnet: a subnet to match
   :param string ip: the IP address or IPaddress port combination to match the subnet to.
+  :param array domains: An array of strings used to fill a :ref:`dns-suffix-match-group`.
 
+If the optional ``domains`` argument is given to this function, only queries for names matching the :ref:`dns-suffix-match-group` will use the value ``M`` to determine the outgoing ECS; other queries will use the value ``S``.
```
[the remark this answers is not in the document]
Comment (Habbie) [reply pdns/recursordist/docs/lua-config/proxymapping.rst:53]: You're right!

**Pull request merged**